### PR TITLE
Newtest

### PR DESCRIPTION
**Pull request by DevloperAmanSingh**
### Summary by CodeRabbit

- New Feature: Introduced an automated code review workflow using GitHub Actions and the `coderabbitai/ai-pr-reviewer` action. This will streamline the code review process by providing AI-powered insights.
- New Feature: Added new protocols, classes, and utility functions to enhance code analysis capabilities. These changes improve the accuracy of code chunking, review findings generation, and pull request analysis.
- New Feature: Implemented `ASTChunker` and `ChunkSelector` classes for more precise code chunking based on language-specific syntax and context. This results in more targeted and relevant feedback during code reviews.

### CODE REVIEW

File: app/services/analysis/review_engine.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+def _prepare_pr_body(raw_body: str | None) -> str | None:
+    if not raw_body:
+        return None
+    cleaned = raw_body.strip()
+    if not cleaned:
+        return None
+    normalized = _normalize_newlines(cleaned)
+    if len(normalized) <= PR_BODY_CHAR_LIMIT:
+        return normalized
+    return _truncate_text(normalized, PR_BODY_CHAR_LIMIT)
```
[the remark this answers is not in the document]
Comment (DevloperAmanSingh) [reply app/services/analysis/review_engine.py:56]: any more changes needed here?

**Comment by DevloperAmanSingh** [comment]
## Review Summary

- Introduce automated code review workflow using GitHub Actions.
- Implement `coderabbitai/ai-pr-reviewer` action for AI-driven insights.
- Add new protocols and classes to enhance code analysis capabilities.
- Implement `ASTChunker` class for improved code chunking based on syntax.
- Introduce `ChunkSelector` class for targeted feedback during reviews.
- Highlight potential integration challenges with existing workflows.
- Emphasize the need for thorough testing of AI-generated insights.
- Ensure alignment of insights with team standards during reviews.

### Release Notes

### Release Notes

- **Automated Code Review Workflow**: Introduced a new workflow using GitHub Actions to streamline the code review process.
- **AI-Driven Insights**: Integrated the `coderabbitai/ai-pr-reviewer` action for enhanced feedback during code reviews.
- **Improved Code Analysis**: Added `ASTChunker` and `ChunkSelector` classes for more accurate code chunking based on language syntax.
- **Targeted Feedback**: Enhanced ability to provide specific, actionable insights during code reviews.
- **Integration Considerations**: Reviewers should be aware of potential challenges with existing workflows and ensure thorough testing.
- **Testing Required**: Emphasis on validating AI-generated insights to align with team standards before full adoption.

File: .github/workflows/ai-pr-reviewer.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: coderabbitai/ai-pr-reviewer@latest
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+        with:
+          debug: false
+          review_simple_changes: false
+          review_comment_lgtm: false
```
Comment (DevloperAmanSingh) [inline .github/workflows/ai-pr-reviewer.yml:29]: ### Review Comments

**1-29:**
- **Explanation:** The workflow is set up to trigger on both `pull_request` and `pull_request_review_comment` events. However, the `pull_request_review_comment` event is not typically used to trigger workflows that perform actions on pull requests. This could lead to unexpected behavior or unnecessary executions of the workflow when comments are made on reviews.
  
  Additionally, the `concurrency` group is using a combination of `github.event.number`, `github.head_ref`, and `github.sha`, which may not always yield a unique identifier for the workflow execution, especially in cases where multiple comments are made on the same pull request.

- **Recommended Fix:**
  Consider removing the `pull_request_review_comment` trigger or ensuring that the workflow is designed to handle it appropriately. Also, revise the concurrency group to ensure it uniquely identifies each workflow run.

  ```diff
  on:
    pull_request:
  # Remove pull_request_review_comment trigger
  # Ensure concurrency group is unique
  concurrency:
    group: ${{ github.repository }}-${{ github.event.number }}-${{ github.workflow }}
  ```

**2-29:**
- **Explanation:** The `with` parameters for the `coderabbitai/ai-pr-reviewer` action include `debug`, `review_simple_changes`, and `review_comment_lgtm`, but there is no documentation or indication of what these parameters do or their expected values. This could lead to confusion for users who are trying to understand or modify the workflow.

- **Recommended Fix:** Add comments to the workflow file explaining the purpose of each parameter and its expected values.

  ```diff
  +        with:
  +          # Set to true to enable debug logging
  +          debug: false
  +          # Set to true to review simple changes
  +          review_simple_changes: false
  +          # Set to true to automatically comment 'LGTM' on reviews
  +          review_comment_lgtm: false
  ```

**3-29:**
- **Explanation:** There is no error handling or fallback mechanism in case the `coderabbitai/ai-pr-reviewer` action fails. If the action fails for any reason (e.g., API key issues, network problems), the workflow will not provide any feedback or logs to help diagnose the issue.

- **Recommended Fix:** Consider adding a step to handle errors gracefully, such as logging the error or sending a notification.

  ```diff
  +      - name: Run AI PR Reviewer
  +        id: ai_review
  +        uses: coderabbitai/ai-pr-reviewer@latest
  +        env:
  +          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  +          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
  +        with:
  +          debug: false
  +          review_simple_changes: false
  +          review_comment_lgtm: false
  +      - name: Check for errors
  +        if: failure()
  +        run: echo "AI PR Reviewer failed. Check the logs for details."
  ```

### Summary
The workflow file has a solid foundation but requires adjustments to improve clarity, error handling, and event handling. Addressing these issues will enhance the reliability and usability of the automated code review process.

File: app/services/agents/review_agent.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            f"(original request {original_start}-{original_end})."
+        )
+        return mapped_start, mapped_end, note
+
+    fallback_start, fallback_end = normalized_patches[0]
+    note = (
+        f"> Note: Comment outside diff; defaulted to hunk [{fallback_start}-{fallback_end}] "
+        f"(original request {original_start}-{original_end})."
+    )
+    return fallback_start, fallback_end, note
```
Comment (DevloperAmanSingh) [inline app/services/agents/review_agent.py:209]: ### Review Comments

**1-209:**
The implementation of the `ReviewAgent` class and its associated methods appears to be well-structured and follows the intended design. However, there are several areas that require attention:

**1-209:**
- **Missing Tests:** There are no tests provided for the `ReviewAgent` class or its methods. Given that this class interacts with external components (like the LLM client), it is crucial to have unit tests that cover various scenarios, including edge cases. Without tests, it is difficult to ensure the reliability of the code.
  
  ```diff
  # Add unit tests for ReviewAgent in a new test file, e.g., test_review_agent.py
  ```

**1-209:**
- **Potential Logic Error in `_normalize_findings`:** The method `_normalize_findings` assumes that the `response` will always contain a "findings" key. If the LLM response structure changes or if the key is missing, it could lead to unexpected behavior. While you have a check for the type of `items`, it would be prudent to also check for the existence of the key itself.

  ```diff
  if "findings" not in response:
      logger.debug("LLM response missing findings key", extra={"file_path": chunk.file_path})
      return findings
  ```

**1-209:**
- **Handling of `prompt_metadata`:** In the `_normalize_findings` method, the handling of `prompt_metadata` could be improved. If `prompt_metadata` is `None`, the code still attempts to access it, which could lead to an `AttributeError`. You should check if `prompt_metadata` is not `None` before trying to access its keys.

  ```diff
  if prompt_metadata is not None and isinstance(prompt_metadata, dict):
      line_ranges = prompt_metadata.get("new_line_ranges", [])
  ```

**1-209:**
- **Type Hinting Consistency:** The type hinting for `line_ranges` in `_normalize_findings` is inconsistent. You initialize it as an empty list but later check if it is a list. It would be clearer to initialize it as an empty list of the expected type.

  ```diff
  line_ranges: list = []
  ```

**1-209:**
- **Error Handling for LLM Responses:** The `review_chunk` method of `LLMClient` is assumed to always return a valid response. It would be beneficial to add error handling for cases where the LLM might return an error or an unexpected format. This could prevent the application from crashing in production.

  ```diff
  if not isinstance(response, dict):
      logger.error("Invalid response from LLM", extra={"file_path": chunk.file_path})
      continue
  ```

**1-209:**
- **Use of `first_chunk`:** The variable `first_chunk` is used to conditionally include PR metadata in the prompt. However, if the `generate_chunks` method returns no chunks, this variable will not be reset, which could lead to incorrect behavior in subsequent calls to `run`. Consider resetting it at the start of the method or ensuring that the method handles empty chunk scenarios gracefully.

  ```diff
  first_chunk = True
  ```

### Summary
Overall, the code is well-structured, but it lacks tests and has some potential logic errors and areas for improvement in error handling and type consistency. Addressing these issues will enhance the robustness and reliability of the `ReviewAgent` class.

File: app/services/chunking/ast_chunker.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                content=snippet,
+                diff_hunk=context.patch,
+                start_line=start_line,
+                end_line=end_line,
+                metadata={
+                    "language": context.language,
+                    "chunk_type": "fallback_window",
+                },
+            )
+        ]
```
Comment (DevloperAmanSingh) [inline app/services/chunking/ast_chunker.py:144]: ### Review Comments

#### 1-144:
The implementation of the `ASTChunker` class appears to be well-structured, but there are a few potential issues and areas for improvement:

1. **Missing Error Handling for Unsupported Languages**:
   The method `_get_parser` does not handle the case where an unsupported language is passed. If `language` is not in `SUPPORTED_LANGUAGES`, `get_language` will raise a `KeyError`. This should be handled gracefully.

   ```diff
   def _get_parser(self, language: str) -> Parser:
       if language not in SUPPORTED_LANGUAGES:
           raise ValueError(f"Unsupported language: {language}")
   ```

2. **Potential for Empty `changed_lines`**:
   In the `generate_chunks` method, if `changed_lines` is empty, the `_fallback_chunks` method is called. However, if `changed_lines` is `None`, it will be converted to an empty set, which is fine, but the logic could be clearer. Consider explicitly checking for `None` before proceeding.

   ```diff
   if changed_lines is None:
       changed_lines = set()
   ```

3. **Inefficient Line Range Calculation**:
   The calculation of `start_line` and `end_line` in `_fallback_chunks` could lead to an IndexError if `changed_lines` is empty. You should check if `changed_lines` is empty before trying to access its minimum or maximum.

   ```diff
   if not changed_lines:
       return []
   ```

4. **Lack of Unit Tests**:
   There are no tests provided for this class. It is crucial to have unit tests that cover various scenarios, including:
   - Supported and unsupported languages.
   - Cases with no changed lines.
   - Cases where the source code is empty.
   - Validating the output of `generate_chunks` against expected `CodeChunk` instances.

   ```python
   def test_generate_chunks_with_supported_language():
       # Add test implementation here

   def test_generate_chunks_with_unsupported_language():
       # Add test implementation here

   def test_generate_chunks_with_no_changed_lines():
       # Add test implementation here
   ```

5. **Type Hinting for `_node_intersects`**:
   The `_node_intersects` method lacks type hints for its parameters. Adding type hints improves code readability and helps with static analysis.

   ```diff
   @staticmethod
   def _node_intersects(node: Node, changed_lines: Set[int]) -> bool:
   ```

6. **Use of `continue` Statement**:
   The use of `continue` after appending to `chunks` is unnecessary since it is the last statement in the loop. This can be removed for clarity.

   ```diff
   # Remove the continue statement
   ```

### Summary
Overall, the `ASTChunker` class is a solid implementation, but it requires some error handling, clarity improvements, and the addition of unit tests to ensure robustness. Addressing these issues will enhance the reliability and maintainability of the code.
